File: admin_web/src/types/monitoringZone.ts
export type MonitoringZoneShape = 'circle' | 'polygon';

export interface MonitoringZoneCoordinate {
  lat: number;
  lng: number;
}

export interface MonitoringZone {
  id: string;
  name: string;
  description: string;
  shape: MonitoringZoneShape;
  center_lat: number;
  center_lng: number;
  radius_meters: number;
  polygon_points: MonitoringZoneCoordinate[] | null;
  people_count: number;
  created_at: string;
}

export type ZoneTransitionEventType = 'entered' | 'exited';
export type ZoneTransitionSource = 'foreground' | 'background';
export type ZoneTransitionDeliveryStatus = 'queued' | 'sent' | 'failed' | 'skipped';

export interface ZoneTransitionEvent {
  id: string;
  zone_id: string;
  zone_name: string;
  user_id: string;
  device_id: string;
  event_type: ZoneTransitionEventType;
  occurred_at: string;
  latitude: number;
  longitude: number;
  source: ZoneTransitionSource;
  delivery_status: ZoneTransitionDeliveryStatus;
  notification_outbox_id: string | null;
}

export interface LivePresenceCell {
  cell_key: string;
  center_lat: number;
  center_lng: number;
  people_count: number;
  latest_observed_at: string;
}

export interface LiveLocationSession {
  id: string;
  started_by: string;
  incident_id: string;
  reason: string;
  status: 'active' | 'ended' | 'expired';
  started_at: string;
  expires_at: string;
  ended_at: string | null;
}

export interface LiveExactPin {
  device_id: string;
  user_id: string;
  latitude: number;
  longitude: number;
  observed_at: string;
  source: ZoneTransitionSource;
  accuracy_meters: number | null;
}

export interface LiveLocationAuditEntry {
  id: string;
  actor_user_id: string;
  action: string;
  session_id: string | null;
  incident_id: string | null;
  details: Record<string, unknown>;
  created_at: string;
}

File: mobile_app/lib/models/monitoring_zone.dart
class MonitoringZoneCoordinate {
  final double lat;
  final double lng;

  const MonitoringZoneCoordinate({
    required this.lat,
    required this.lng,
  });
}

class MonitoringZone {
  final String id;
  final String name;
  final String description;
  final String shape; // circle | polygon
  final double centerLat;
  final double centerLng;
  final double radiusMeters;
  final List<MonitoringZoneCoordinate> polygonPoints;
  final int peopleCount;
  final DateTime createdAt;

  const MonitoringZone({
    required this.id,
    required this.name,
    required this.description,
    required this.shape,
    required this.centerLat,
    required this.centerLng,
    required this.radiusMeters,
    required this.polygonPoints,
    required this.peopleCount,
    required this.createdAt,
  });

  bool get isPolygon => shape == 'polygon' && polygonPoints.length >= 3;

  bool get isCircle => shape != 'polygon';

  factory MonitoringZone.fromJson(Map<String, dynamic> json) {
    final rawPoints = json['polygon_points'];
    final points = <MonitoringZoneCoordinate>[];

    if (rawPoints is List) {
      for (final item in rawPoints) {
        if (item is! Map) continue;
        final lat = double.tryParse(item['lat']?.toString() ?? '');
        final lng = double.tryParse(item['lng']?.toString() ?? '');
        if (lat == null || lng == null) continue;
        points.add(MonitoringZoneCoordinate(lat: lat, lng: lng));
      }
    }

    return MonitoringZone(
      id: json['id']?.toString() ?? '',
      name: (json['name'] as String?)?.trim().isNotEmpty == true
          ? json['name'] as String
          : ((json['shape']?.toString() == 'polygon') ? 'Polygon Zone' : 'Monitoring Zone'),
      description: (json['description'] as String?)?.trim() ?? '',
      shape: json['shape']?.toString() == 'polygon' ? 'polygon' : 'circle',
      centerLat: (json['center_lat'] as num?)?.toDouble() ?? 0,
      centerLng: (json['center_lng'] as num?)?.toDouble() ?? 0,
      radiusMeters: (json['radius_meters'] as num?)?.toDouble() ?? 0,
      polygonPoints: points,
      peopleCount: (json['people_count'] as num?)?.toInt() ?? 0,
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ?? DateTime.now(),
    );
  }
}
